In [51]:
import pandas as pd
import numpy as np
import datetime as dt

In [3]:
df = pd.read_csv('test_data.csv', index_col=0, parse_dates=['creation', 'deadline'])

In [4]:
df.head()

user_id  order_id  type_id  category_id  title  description  \
0    17251   17250.0      1.0         16.0   14.0         53.0   
1    12299       NaN      1.0         13.0   24.0          0.0   
2      720     624.0     15.0        176.0   15.0          0.0   
3     6187    6119.0      NaN          1.0   41.0          0.0   
4    21360   21372.0      1.0          1.0   48.0          0.0   

             creation            deadline  wordsFrom  wordsTo  source_of_user  \
0 2015-10-01 21:01:44 2015-10-23 09:04:47      275.0    275.0             NaN   
1 2015-09-24 22:11:55 2015-10-02 09:04:47     1100.0   2200.0             5.0   
2 2015-09-03 12:11:01 2015-09-10 09:04:47      825.0   1650.0             5.0   
3 2015-09-15 16:46:20 2015-09-23 09:04:47     1100.0      NaN             0.0   
4 2015-10-07 15:09:01 2015-10-09 09:04:47      300.0    300.0             1.0   

   is_paid  is_test  file_count  
0      0.0      1.0         0.0  
1      0.0      0.0         0.0  
2      0.0      0.0         0.0  
3      0.0      0.0         0.0  
4      0.0      0.0         0.0

In [5]:
df.shape[0]

13425

In [6]:
df.dtypes

user_id                    int64
order_id                 float64
type_id                  float64
category_id              float64
title                    float64
description              float64
creation          datetime64[ns]
deadline          datetime64[ns]
wordsFrom                float64
wordsTo                  float64
source_of_user           float64
is_paid                  float64
is_test                  float64
file_count               float64
dtype: object

In [7]:
df.describe()

user_id      order_id       type_id   category_id         title  \
count  13425.000000  12626.000000  12926.000000  13056.000000  12675.000000   
mean   17219.339516  17272.053224      7.633220     51.373162     32.929073   
std     9911.980823   9956.447294     12.042857     73.304662     27.153418   
min        3.000000      0.000000      1.000000      1.000000      1.000000   
25%     8614.000000   8625.750000      1.000000      1.000000     14.000000   
50%    17181.000000  17236.500000      1.000000     19.000000     23.000000   
75%    25784.000000  25855.750000     12.000000     42.000000     44.000000   
max    34344.000000  34467.000000     56.000000    248.000000    110.000000   

        description      wordsFrom        wordsTo  source_of_user  \
count  13035.000000   13255.000000   12913.000000    12809.000000   
mean      78.781741    1127.210637    1568.944242        2.779998   
std      415.599704    1863.373746    1997.703739        2.033027   
min        0.000000       0.000000       0.000000        0.000000   
25%        0.000000     550.000000     550.000000        1.000000   
50%        0.000000    1100.000000    1650.000000        2.000000   
75%        0.000000    1100.000000    2200.000000        5.000000   
max     9484.000000  100000.000000  100000.000000        5.000000   

            is_paid       is_test    file_count  
count  12964.000000  13030.000000  12797.000000  
mean       0.093490      0.024021      0.088224  
std        0.291129      0.153122      0.571915  
min        0.000000      0.000000      0.000000  
25%        0.000000      0.000000      0.000000  
50%        0.000000      0.000000      0.000000  
75%        0.000000      0.000000      0.000000  
max        1.000000      1.000000     23.000000

In [8]:
df.isna().sum()

user_id             0
order_id          799
type_id           499
category_id       369
title             750
description       390
creation          426
deadline          775
wordsFrom         170
wordsTo           512
source_of_user    616
is_paid           461
is_test           395
file_count        628
dtype: int64

In [9]:
df.nunique()

user_id           9984
order_id          9571
type_id             54
category_id        163
title              110
description        728
creation          9767
deadline          1160
wordsFrom          150
wordsTo            150
source_of_user       6
is_paid              2
is_test              2
file_count          14
dtype: int64

In [10]:
# удаление полных дублей
df = df.drop_duplicates()

In [11]:
# удаление полей, где значение is_paid не определено, их невозможно интерпретировать
df = df.dropna(subset = ['is_paid'])

In [12]:
# сортировка по всем колонкам в датафрейме
df = df.sort_values(df.columns.tolist())

In [13]:
# удаление оставшихся дублей
df = df.drop_duplicates(subset = ['order_id'])
df = df.drop([1189])

In [14]:
df.shape[0]

9333

In [15]:
df.describe()

user_id      order_id      type_id  category_id        title  \
count   9333.000000   9333.000000  9078.000000  9124.000000  8920.000000   
mean   17203.185257  17288.309868     7.666226    51.464051    33.101682   
std     9926.627287   9957.245805    12.150704    73.512249    27.221863   
min        3.000000      0.000000     1.000000     1.000000     1.000000   
25%     8538.000000   8622.000000     1.000000     1.000000    14.000000   
50%    17182.000000  17262.000000     1.000000    19.000000    24.000000   
75%    25825.000000  25930.000000    12.000000    42.000000    44.000000   
max    34344.000000  34467.000000    56.000000   248.000000   110.000000   

       description      wordsFrom        wordsTo  source_of_user      is_paid  \
count  9120.000000    9230.000000    9058.000000     8961.000000  9333.000000   
mean     77.229496    1129.464572    1569.972179        2.792657     0.093432   
std     397.769624    1875.292361    1988.301500        2.032129     0.291052   
min       0.000000       0.000000       0.000000        0.000000     0.000000   
25%       0.000000     550.000000     550.000000        1.000000     0.000000   
50%       0.000000    1100.000000    1650.000000        2.000000     0.000000   
75%       0.000000    1100.000000    2200.000000        5.000000     0.000000   
max    9484.000000  100000.000000  100000.000000        5.000000     1.000000   

           is_test   file_count  
count  9087.000000  8934.000000  
mean      0.024431     0.086747  
std       0.154390     0.596247  
min       0.000000     0.000000  
25%       0.000000     0.000000  
50%       0.000000     0.000000  
75%       0.000000     0.000000  
max       1.000000    23.000000

In [16]:
df.isna().sum()

user_id             0
order_id            0
type_id           255
category_id       209
title             413
description       213
creation          222
deadline          442
wordsFrom         103
wordsTo           275
source_of_user    372
is_paid             0
is_test           246
file_count        399
dtype: int64

In [17]:
df.nunique()

user_id           9317
order_id          9333
type_id             54
category_id        161
title              110
description        685
creation          9100
deadline          1111
wordsFrom          142
wordsTo            142
source_of_user       6
is_paid              2
is_test              2
file_count          14
dtype: int64

In [18]:
# создание колонки со сроком выполнения заказа в часах
df['term'] = ((df.deadline - df.creation)/np.timedelta64(1, 'h')).round(0)

In [19]:
# сохранение очищенного датафрейма
df.to_csv('test_data_clear.csv')

In [20]:
df.head()

user_id  order_id  type_id  category_id  title  description  \
3252         3   11273.0      1.0         19.0   34.0        404.0   
11380        6   22238.0     33.0          1.0   42.0        272.0   
3787        12   13400.0      1.0          1.0    7.0          0.0   
1230        13     370.0     45.0          3.0   31.0        213.0   
5343        14   25469.0     16.0          2.0    5.0         19.0   

                 creation            deadline  wordsFrom  wordsTo  \
3252  2015-09-23 12:45:46 2015-09-26 16:04:48     1100.0   1100.0   
11380 2015-10-08 17:02:37 2015-10-10 16:04:48     1800.0   1800.0   
3787  2015-09-26 23:02:46 2015-10-13 17:04:47      275.0    275.0   
1230  2015-09-02 16:27:46 2015-09-08 17:04:47     2750.0   2750.0   
5343  2015-10-13 17:19:11 2015-10-20 17:04:47     1650.0   1650.0   

       source_of_user  is_paid  is_test  file_count   term  
3252              0.0      1.0      0.0         0.0   75.0  
11380             0.0      1.0      0.0         NaN   47.0  
3787              5.0      1.0      0.0         0.0  402.0  
1230              2.0      0.0      0.0         4.0  145.0  
5343              5.0      0.0      0.0         0.0  168.0

Дальнейшие расчеты и построение визуализации были проведены в Tableau

Ссылка на аналитический дашборд - https://public.tableau.com/app/profile/anton.milchakov/viz/_16361388804460/Dashboard1

Были проверены гипотезы по влиянию следующих полей заказа на конверсию в оплату:

 - title - длина названия заказа
 - description - длина описания заказа
 - term (deadline - creation) - срок выполнения заказа
 - file_count - кол-во приклепленных файлов к заказу за первые 10 минут
 - wordsFrom - нижняя граница слов в выполненном заказе
 - wordsTo - верхняя граница слов в выполненном заказе

### Выводы:

##### Общее:
 - После очистки исходных данных осталось 9333 уникальных заказа с заполненным полем is_paid в период с 01.09.2015 по 27.10.2015
 - Из них оплачено - 872
 - Конверсия в оплату - 9.34%

##### Поле title:
 - заказы с очень коротким (длина до 10) и очень длинным (длина более 60) названием имеют конверсию ниже чем заказы с длиной от 11 до 60

##### Поле description:
 - прямая коррелляция конверсии и длины описания - чем длиннее описание тем выше конверсия
 - заказы с незаполненым полем имеют наименьший показатель

##### Поля deadline и creation:
 - была проведена оценка влияния срока выполнения заказа (term) на конверсию
 - низкие показатели у срочных заказов (до 8 часов) и заказов с большим сроком (более 6 суток)
 - наибольшая конверсия у заказов, которые необходимо выполнить в течение 3-6 суток
 
##### Полe file_count:
 - наличие прикрепленных файлов оказывает значительное влияние на конверсию (39.55% - есть прикрепленные файлы, 8.04 % - нет)
 
##### Поля wordsFrom и wordsTo:
 - значения в полях зависят от типа заказа, поэтому нет отчетливой тенденции.
 - возможно, стоит рассмотреть возможность удаления этих полей из формы заказа, т.к. об объеме работы заказчик может сообщить в описании или в переписке с исполнителем.

 
В дашборде есть фильтры, где можно посмотреть на показатели конверсии в различных разрезах:
 - type_id - тип заказа
 - category_id - предмет заказа
 - source_of_user - источник пользователя
 - is_test - тестовый ли пользователь
 
В целом, по различным разрезам тенденции сохраняются, за исключением позиций с небольшим количеством заказов

Ниже топы по количеству заказов в различных разрезах:

In [47]:
df.groupby('type_id') \
  .agg({'order_id': 'count'}) \
  .sort_values('order_id', ascending=False) \
  .head()

order_id
type_id          
1.0          6093
12.0          621
16.0          504
27.0          134
33.0          130

In [48]:
df.groupby('category_id') \
  .agg({'order_id': 'count'}) \
  .sort_values('order_id', ascending=False) \
  .head()

order_id
category_id          
1.0              2749
19.0              996
42.0              505
2.0               296
13.0              232

In [50]:
df.groupby('source_of_user') \
  .agg({'order_id': 'count'}) \
  .sort_values('order_id', ascending=False)

order_id
source_of_user          
5.0                 3986
1.0                 3703
0.0                  692
2.0                  375
3.0                  178
4.0                   27